cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -261,6 +262,95 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
     TArray<FPType, cpu> normBufferB;
 };
 
+template <typename FPType, CpuType cpu>
+bool solveEquationsSystemWithCholesky(FPType * a, FPType * b, size_t n, size_t nX, bool sequential)
+{
+    /* POTRF and POTRS parameters */
+    char uplo     = 'U';
+    DAAL_INT info = 0;
+
+    /* Perform L*L' decomposition of A */
+    if (sequential)
+    {
+        Lapack<FPType, cpu>::xxpotrf(&uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, &info);
+    }
+    else
+    {
+        Lapack<FPType, cpu>::xpotrf(&uplo, (DAAL_INT *)&n, a, (DAAL_INT *)&n, &info);
+    }
+    if (info != 0) return false;
+
+    /* Solve L*L' * x = b */
+    if (sequential)
+    {
+        Lapack<FPType, cpu>::xxpotrs(&uplo, (DAAL_INT *)&n, (DAAL_INT *)&nX, a, (DAAL_INT *)&n, b, (DAAL_INT *)&n, &info);
+    }
+    else
+    {
+        Lapack<FPType, cpu>::xpotrs(&uplo, (DAAL_INT *)&n, (DAAL_INT *)&nX, a, (DAAL_INT *)&n, b, (DAAL_INT *)&n, &info);
+    }
+    return (info == 0);
+}
+
+template <typename FPType, CpuType cpu>
+bool solveEquationsSystemWithPLU(FPType * a, FPType * b, size_t n, size_t nX, bool sequential)
+{
+    /* GETRF and GETRS parameters */
+    char trans    = 'N';
+    DAAL_INT info = 0;
+
+    TArray<DAAL_INT> ipiv(n);
+
+    /* Perform P*L*U decomposition of A */
+    if (sequential)
+    {
+        Lapack<FPType, cpu>::xxgetrf((DAAL_INT *)&n, (DAAL_INT *)&n, a, (DAAL_INT *)&n, ipiv, &info);
+    }
+    else
+    {
+        Lapack<FPType, cpu>::xgetrf((DAAL_INT *)&n, (DAAL_INT *)&n, a, (DAAL_INT *)&n, ipiv, &info);
+    }
+    if (info != 0) return false;
+
+    /* Solve P*L*U * x = b */
+    if (sequential)
+    {
+        Lapack<FPType, cpu>::xxgetrs(&trans, (DAAL_INT *)&n, (DAAL_INT *)&nX, a, (DAAL_INT *)&n, ipiv, b, (DAAL_INT *)&n, &info);
+    }
+    else
+    {
+        Lapack<FPType, cpu>::xgetrs(&trans, (DAAL_INT *)&n, (DAAL_INT *)&nX, a, (DAAL_INT *)&n, ipiv, b, (DAAL_INT *)&n, &info);
+    }
+    return (info == 0);
+}
+
+template <typename FPType, CpuType cpu>
+bool solveEquationsSystem(FPType * a, FPType * b, size_t n, size_t nX, bool sequential)
+{
+    TArray<FPType, cpu> aCopy(n * n);
+    TArray<FPType, cpu> bCopy(n);
+
+    int copy_status = services::internal::daal_memcpy_s(aCopy.get(), n * n * sizeof(FPType), a, n * n * sizeof(FPType));
+    copy_status |= services::internal::daal_memcpy_s(bCopy.get(), n * sizeof(FPType), b, n * sizeof(FPType));
+
+    if (copy_status != 0)
+    {
+        return false;
+    }
+
+    if (!solveEquationsSystemWithCholesky<FPType, cpu>(a, b, n, nX, sequential))
```

Comment:
Can we use LUP always? Or there are perf improvement for cholesky?

Reply:
It was found out that Cholesky is actually faster for ALS because of more intensive equations system solver usage